torch/csrc/jit/codegen/cuda/ir_builder.cpp

```diff
@@ -30,13 +30,21 @@ Val* IrBuilder::newArithmeticExpr(BinaryOpType op_type, Val* lhs, Val* rhs) {
   TORCH_CHECK(
       lhs != nullptr && rhs != nullptr,
       "Either lhs or rhs is a nullptr in newArithmeticExpr.");
-  TORCH_CHECK(
-      lhs->dtype() == rhs->dtype(),
-      "Incompatible operand types: ",
-      lhs->dtype(),
-      " and ",
-      rhs->dtype());
-  auto result = newScalar(lhs->dtype());
+  auto dtype = lhs->dtype();
+  if (lhs->dtype() != rhs->dtype()) {
+    if ((lhs->dtype() == DataType::Int && rhs->dtype() == DataType::Int32) ||
```

Comment:
nit: leave a comment about why only enabling type promotion for integers but not other types, a pointer to the issue should be sufficient.

Reply:
In principle, I think we should keep these `IrBuilder` functions as simple as possible since they are just used by the lowering for scalar computations. If we could enforce strict typing with no implicit type promotion, I think we should do so. However, for int and int64_t, our usages are pretty loose. Originally we only had int64_t, then we added nvfuser_index_t and replaced the types of some of the values from int64_t to int just at the beginning of lowering. This results in inconsistent usage of integer types, e.g., here: https://github.com/csarofeen/pytorch/blob/devel/torch/csrc/jit/codegen/cuda/index_compute.cpp#L1532

```
std::stringstream ss;
      ss << "T" << producer_tv->name() << ".stride[" << stride_i++ << "]";
      strides[i] =
          SimplifyingIrBuilder::create<NamedScalar>(ss.str(), DataType::Int);
```

This is kind of cheating the type system as its real type is nvfuser_index_t, which can be int, but we are keeping it int64_t. The index part of the expression is also int64_t, and so `IrBuilder::mulExpr` doesn't complain. Similarly, this is not strictly correct since `threadIdx.x` should be int not int64_t.

```
NamedScalar* NamedScalar::getParallelIndex(ParallelType p_type) {
  TORCH_INTERNAL_ASSERT(FusionGuard::getCurFusion() != nullptr);
  std::string parallel_ind = stringifyThread(p_type);
  return IrBuilder::create<NamedScalar>(parallel_ind, DataType::Int);
}
```

`zeroVal()` and `oneVal()` are also int64_t, which should probably be int.

We could try to make everything more consistent, but at this point the easiest workaround is to let the type check a little loose on integer types. As this typing issue is only with integer types, I added the promotion only for int and int64_t.

Reply:
I see, it makes sense.

Reply:
Will add more comments to the code